csrc/scheduler/registry.cpp

```diff
@@ -63,17 +63,43 @@ bool rejectScheduleFusionInputRequirement(
   return false;
 }
 
-bool rejectScheduleForSelectLikeOps(
+// TODO: remove this requirement entirely
+bool rejectScheduleForMemoryPromotion(
     Fusion* fusion,
     ScheduleHeuristic schedule_strategy) {
   for (auto expr : fusion->exprs()) {
-    // For now, only relax the input requirement with take_along_axis.
-    // TODO: remove this requirement entirely
-    if ((expr->isOneOf<SelectOp, IndexSelectOp>() ||
-         (expr->isA<TorchGatherOp>() &&
-          !expr->as<TorchGatherOp>()->exactSizes())) &&
-        rejectScheduleFusionInputRequirement(expr, schedule_strategy)) {
-      return true;
+    if (expr->isOneOf<SelectOp, IndexSelectOp, TorchGatherOp>()) {
+      // For now, only relax the input requirement when it's
+      // take_along_axis. Also since this would require memory
+      // promotion, i.e., persistent global sync in the case of
+      // block-parallel ops, it needs to be explictly enabled unless
+      // it's for the persistent scheuduler, which is already
+      // persistent. For block-parallel ops, it means the problem size
+      // is small enough to fit in registers, so the required capacity
+      // of the shared memory is likely small enough.
+      if (expr->isA<TorchGatherOp>() &&
+          expr->as<TorchGatherOp>()->exactSizes() &&
+          (isOptionEnabled(EnableOption::MemoryPromotion) ||
+           schedule_strategy == ScheduleHeuristic::Persistent)) {
```

Comment:
I don't know how our persistent kernel is implemented, but does persistent kernel implies cooperative launch? For example, if I have a tensor of shape (100000000, 1024), and I want to do a normalization on the second dim, can I just launch 100000000 blocks and each block do block reduction and broadcasting? For this case, I don't think we have block sync, but if we have `take_along_axis` along the first dim, does the memory promotion create block sync, which then will fail?

Reply:
Oh, I'm sorry but that was just I forgot removing a temporary change I was playing with. `schedule_strategy == ScheduleHeuristic::Persistent` should be removed. I'll update the PR and rerun the tests.

Reply:
Updated